kernel/src/schema_compat.rs

```diff
@@ -0,0 +1,331 @@
+        );
+
+        let read_field_names: HashSet<String> =
+            read_type.fields.keys().map(|x| x.to_lowercase()).collect();
+        require!(
+            read_field_names.len() == read_type.fields.len(),
+            Error::generic("Delta tables don't allow field names that only differ by case")
+        );
+
+        // Check that the field names are a subset of the read fields.
+        if !field_map.keys().all(|name| read_field_names.contains(name)) {
+            return Err(Error::generic(
+                "Struct has column that does not exist in the read schema",
+            ));
+        }
+        for read_field in read_type.fields() {
+            match field_map.get(&read_field.name().to_lowercase()) {
+                Some(existing_field) => existing_field.can_read_as(read_field)?,
+                None => {
+                    // Note: Delta spark does not perform the following check. Hence it ignores fields
+                    // that exist in the read schema that aren't in this schema.
+                    require!(
+                        read_field.is_nullable(),
+                        Error::generic(
+                            "read type has non-nullable column that does not exist in this struct",
+                        )
+                    );
+                }
+            }
+        }
+        Ok(())
+    }
+}
+
+impl crate::schema::DataType {
+    /// Returns `Ok` if this [`DataType`] can be read as `read_type` in the read schema.
+    fn can_read_as(&self, read_type: &Self) -> DeltaResult<()> {
+        match (self, read_type) {
+            (Self::Array(self_array), Self::Array(read_array)) => {
+                NullabilityFlag(self_array.contains_null())
+                    .can_read_as(NullabilityFlag(read_array.contains_null()))?;
+                self_array
+                    .element_type()
+                    .can_read_as(read_array.element_type())?;
+            }
+            (Self::Struct(self_struct), Self::Struct(read_struct)) => {
+                self_struct.can_read_as(read_struct)?
+            }
+            (Self::Map(self_map), Self::Map(read_map)) => {
+                NullabilityFlag(self_map.value_contains_null())
+                    .can_read_as(NullabilityFlag(read_map.value_contains_null()))?;
+                self_map.key_type().can_read_as(read_map.key_type())?;
+                self_map.value_type().can_read_as(read_map.value_type())?;
+            }
+            (a, b) => {
+                // TODO: In the future, we will change this to support type widening.
+                // See: https://github.com/delta-io/delta-kernel-rs/issues/623
+                require!(
+                    a == b,
+                    Error::generic(format!("Types {} and {} are not compatible", a, b))
+                );
+            }
+        };
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
if there are some existing useful tests in delta-spark perhaps we can have an issue to evaluate/migrate those?

Reply:
Created issue: #655